spec/versionedsubmitblockrequest.go

```diff
@@ -0,0 +1,531 @@
+	case consensusspec.DataVersionCapella:
+		if v.Capella == nil {
+			return consensusbellatrix.ExecutionAddress{}, errors.New("no data")
+		}
+		if v.Capella.Message == nil {
+			return consensusbellatrix.ExecutionAddress{}, errors.New("no data message")
+		}
+		return v.Capella.Message.ProposerFeeRecipient, nil
+	default:
+		return consensusbellatrix.ExecutionAddress{}, errors.New("unsupported version")
+	}
+}
+
+// ProposerPubkey returns the proposer fee recipient of the request.
+func (v *VersionedSubmitBlockRequest) ProposerPubkey() (phase0.BLSPubKey, error) {
+	if v == nil {
+		return phase0.BLSPubKey{}, errors.New("nil struct")
+	}
+	switch v.Version {
+	case consensusspec.DataVersionBellatrix:
+		if v.Bellatrix == nil {
+			return phase0.BLSPubKey{}, errors.New("no data")
+		}
+		if v.Bellatrix.Message == nil {
+			return phase0.BLSPubKey{}, errors.New("no data message")
+		}
+		return v.Bellatrix.Message.ProposerPubkey, nil
+	case consensusspec.DataVersionCapella:
+		if v.Capella == nil {
+			return phase0.BLSPubKey{}, errors.New("no data")
+		}
+		if v.Capella.Message == nil {
+			return phase0.BLSPubKey{}, errors.New("no data message")
+		}
+		return v.Capella.Message.ProposerPubkey, nil
+	default:
+		return phase0.BLSPubKey{}, errors.New("unsupported version")
+	}
+}
+
+// ParentHash returns the parent hash of the request.
+func (v *VersionedSubmitBlockRequest) ParentHash() (phase0.Hash32, error) {
+	if v == nil {
+		return phase0.Hash32{}, errors.New("nil struct")
+	}
+	switch v.Version {
+	case consensusspec.DataVersionBellatrix:
+		if v.Bellatrix == nil {
+			return phase0.Hash32{}, errors.New("no data")
+		}
+		if v.Bellatrix.Message == nil {
+			return phase0.Hash32{}, errors.New("no data message")
+		}
+		return v.Bellatrix.Message.ParentHash, nil
+	case consensusspec.DataVersionCapella:
+		if v.Capella == nil {
+			return phase0.Hash32{}, errors.New("no data")
+		}
+		if v.Capella.Message == nil {
+			return phase0.Hash32{}, errors.New("no data message")
+		}
+		return v.Capella.Message.ParentHash, nil
+	default:
+		return phase0.Hash32{}, errors.New("unsupported version")
+	}
+}
+
+// Value returns the value of the request.
+func (v *VersionedSubmitBlockRequest) Value() (*uint256.Int, error) {
+	if v == nil {
+		return nil, errors.New("nil struct")
+	}
+	switch v.Version {
+	case consensusspec.DataVersionBellatrix:
+		if v.Bellatrix == nil {
+			return nil, errors.New("no data")
+		}
+		if v.Bellatrix.Message == nil {
+			return nil, errors.New("no data message")
+		}
+		return v.Bellatrix.Message.Value, nil
+	case consensusspec.DataVersionCapella:
+		if v.Capella == nil {
+			return nil, errors.New("no data")
+		}
+		if v.Capella.Message == nil {
+			return nil, errors.New("no data message")
+		}
+		return v.Capella.Message.Value, nil
+	default:
+		return nil, errors.New("unsupported version")
+	}
+}
+
+// MessageHashTreeRoot returns the hash tree root of the message of the bid.
```

Comment:
Not a fan of this, I'd prefer to see a call for `BidTrace()` that returned a `BidTrace` (or `VersionedBidTrace` if you want to future-proof the call) and then users can call `HashTreeRoot()` on that.

Reply:
I agree, however was following the pattern of other structs such as `VersionedSignedBlindedBeaconBlock()` where the `Root()` function does the same thing